### PR TITLE
GameINI: Add some speed hacks

### PR DESCRIPTION
**Pull request by SuperSamus**
Add patches to improve performance on some games.

### Idle loop detection patches
Avoid accurately emulating the game twiddling its thumbs, removing the need to lower the Emulated CPU Clock for performance.

The list of known games with an undetected idle loop is the following. The ones with a patch added by the PR have a check.
- [x] Need for Speed: Nitro
- [ ] (Almost?) all Need for Speed games by Black Box
- [ ] Excite Truck (normal gameplay)
  - Uncapped during main menu and slow-motion caused by car crash (where it enables interpolated rendering)
  - Fixed by #14637.
- [ ] Pokémon Colosseum
  - Fixed by #14636.
- [ ] Knights of the Temple: Infernal Crusade
  - Fixed by #14638.
- [x] Rayman: Raving Rabbids (patch only for gameplay, FMVs are not patched)
- [ ] Skylanders: Swap Force
  - Fixed by #14717.
- [ ] Games by Retro Studios
  - Fixed by #14682.

### Limit internal frame rate patches:
The following games internally run at an uncapped frame rate, making them challenging to emulate since Dolphin will process all these frames that then go to waste (they are only shown with Immediately Present XFB). Plus, their animation delta is inconsistent, and VBI Skip does nothing on them. 
The patch limits their frame rate to the VBI frequency, making them behave like a normal game, and removing the need to lower the Emulated CPU Clock for performance.

While capping the frame rate makes input lag worse, for the "high FPS only for better input lag" purposes there is still VBI Frequency Override.

The list of known games with an uncapped frame rate is the following.
- [x] 007: Quantum of Solace (only the intro)
  - Fixed by #14300.
- [x] 4x4 EVO 2
  - Fixed by #14295.
- [x] Alien Hominid
- [ ] All Star Cheer Squad
  - Probably All Star Cheer Squad 2 (Same developer.)
- [x] Bully: Scholarship Edition (loading screens)
  - Works around hangs during loading that still happened even with Emulate Disc Speed off (still, https://bugs.dolphin-emu.org/issues/8279 can't be considered fixed).
- [ ] Cake Mania: In the Mix!
- [ ] Crash Tag Team Racing (initial screen only)
- [x] Conduit 2
  - Fixes https://bugs.dolphin-emu.org/issues/7132
- [x] Dark Summit
  - Fixed by #14301.
- [x] Driver: San Francisco (gameplay only)
  - Fixed by #14102.
- [ ] Emergency Heroes
- [ ] Heavy Fire: Black Arms
- [ ] Heavy Fire: Special Operations
- [x] Hulk
  - The game has audio and gameplay issues at high frame rates (even on the PC version). Thus, capping the frame rate fixes https://bugs.dolphin-emu.org/issues/13107.
- [ ] Midway Arcade Treasures 3 (intro FMVs only)
- [x] Monster 4x4: Stunt Racer (gameplay only)
  - Fixed by #14297.
- [x] Monsters, Inc. Scream Arena
- [ ] My Fitness Coach
- [x] Need for Speed: Hot Pursuit 2
  - Fixed by #14279.
- [ ] Pirates vs Ninjas Dodgeball
- [ ] Scarface: The World is Yours (intro only)
- [ ] Rabbids Go Home
  - Some regions fixed by the Rabbids Lab PR and #14456.
- [x] Rabbids Lab
  - Fixed by #14280.
- [ ] Ratatouille (FMVs and loading screens)
- [ ] Rec Room Games (credits only)
- [x] The Simpsons - Hit & Run (intro only)
- [x] Tetris Worlds
- [ ] Worms: Battle Islands (loading screens)
- [ ] Xenoblade Chronicles (only the very first loading on boot)
- [ ] Your Shape

### Credits
- The PAL patches for Need for Speed: Nitro, Rayman: Raving Rabbids, Monster Inc. Scream Arena and Tetris Worlds me.
- Revision 0 NTSC of Tetris Worlds corrected by @Tilka.
- Alien Hominid's patch by [an unknown author](https://wiki.dolphin-emu.org/index.php?title=Alien_Hominid&diff=141508&oldid=133744).
- The ports for other regions, and all the other games by @JMC47 (with my guidance).

### CODE REVIEW

File: Data/Sys/GameSettings/R7XP69.ini [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# R7XP69 - Need for Speed: Nitro
+
+[OnFrame]
+# Changes an idle loop that Dolphin is unable to detect, to one that it can detect.
+#
+# The loop waits for a vblank interrupt, spamming `OSYieldThread` in the meantime.
+# The patch replaces the call to `VIWaitForRetrace, which still executes the other threads,
```
Comment (Rexogamer) [inline Data/Sys/GameSettings/R7XP69.ini]:
```suggestion
# The patch replaces the call to `VIWaitForRetrace`, which still executes the other threads,
```

File: Data/Sys/GameSettings/R7XE69.ini [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# R7XE69 - Need for Speed: Nitro
+
+[OnFrame]
+# Changes an idle loop that Dolphin is unable to detect, to one that it can detect.
+#
+# The loop waits for a vblank interrupt, spamming `OSYieldThread` in the meantime.
+# The patch replaces the call to `VIWaitForRetrace, which still executes the other threads,
```
Comment (Rexogamer) [inline Data/Sys/GameSettings/R7XE69.ini]:
```suggestion
# The patch replaces the call to `VIWaitForRetrace`, which still executes the other threads,
```

**Review by Rexogamer — CHANGES_REQUESTED** [review]

**Comment by JMC47** [comment]
I tried to make a patch for Tetris Worlds, but I'm unable to determine a performance boost.  I put a breakpoint on GXDrawDone and then stepped out and nopped the address.  The code looked like this:

[OnFrame]
$Limit internal FPS to VPS (improves performance)
0x80101B04:dword:0x60000000

**Comment by SuperSamus** [comment]
Codes for the other versions (added in the last push) ported by @JMC47.

One last thing that could be argued is whether these should be enabled by default.

**Comment by SuperSamus** [comment]
Should I add these games to the RetroAchievements allowlist, even though these games have (currently) no achievements?

**Comment by JosJuice** [comment]
So far, the approach has been to add to the allowlist regardless of whether the game has achievements.

**Comment by iwubcode** [comment]
Two other games that may have uncapped framerates:

[Bully Scholarship Edition](https://wiki.dolphin-emu.org/index.php?title=Bully:_Scholarship_Edition) - has an incredibly slow startup speed.  Turning down resolution (presumably to reduce XFB overhead) seems to help the title.  @JMC47 suggested we might be seeing issues due to an uncapped framerate (with some emulation issues making things worse).
[Deadly Creatures](https://wiki.dolphin-emu.org/index.php?title=Deadly_Creatures) - at one point this ran fine, so this might be unrelated.  But it too has a very slow startup speed (though not as bad as Bully).

File: Data/Sys/GameSettings/GTRE78.ini [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# GTRE78 - Tetris Worlds
+
+[OnFrame]
+# After `GXDrawDone`, `nop` a branch that skips a call to `VIWaitForRetrace`.
+$Limit internal FPS to VPS (improves performance)
+0x80101B08:dword:0x60000000
```
Comment (Amphitryon0) [inline Data/Sys/GameSettings/GTRE78.ini:6]: Does this work for both revisions?

Comment (SuperSamus) [reply Data/Sys/GameSettings/GTRE78.ini:6]: Good question...
@JMC47, what was your revision?

Comment (Tilka) [reply Data/Sys/GameSettings/GTRE78.ini:6]: JMC's code only works for revision 1. The address for revision 0 is 0x80101A80.

**Comment by Adamillo** [comment]
I did find a speed hack for Epic Mickey 2, but I doubt that nopping an instruction is the best way to skip idle looping, so I hope someone else can make a better implementation of that game's idle loop skip

**Comment by SuperSamus** [comment]
https://discord.com/channels/521709831132807179/822820107788746812/1400587208854863892
Sorry if that last commit isn't completely related...

**Comment by iwubcode** [comment]
I am _beyond excited_ to say that Bully has been solved by @JMC47 and @SuperSamus .  No more waiting on startup or messing with resolution!!!!!!  They also figured out Conduit 2.  The patches will be added to this PR soon.

EDIT: Bully no longer hangs with VBI Frequency Override enabled during loading transitions...

**Comment by JMC47** [comment]
Once we have all the games added that we want to here, along with an approval from maintainers that this is an appropriate thing to enable by default, I'm good with it.  I've tested most of these and had others test the versions I'm missing.

**Comment by SuperSamus** [comment]
Added more games. Details in the main post.

Drafting again due to RetroAchievement hashes not matching.
I'll undraft when everything else has been approved.

~~EDIT: Also found out that Rayman Raving Rabbids has multiple revisions. Currently missing rev 0 NTSC.~~

The thing to discuss is what name should the patches have, and whether they should be enabled by default.
Heck, not all games here are equal.
Need For Speed: Nitro's idle loop is too complex to ever be detected by Dolphin, while Rayman Raving Rabbids could be detected in the future (it's still far from trivial).
Capping the frame rate for Bully and Hulk fixes severe issues, while for others it's mostly beneficial only for performance.

Arguments for enabling by default:
- Better UX for the user, who won't wonder why a simple game like Tetris Worlds is so demanding to emulate
  - Even better UX if the game has severe issues without the patch
- None of these patches have negative side effects
- While an argument against patches is that they deter from improving Dolphin at its core, it really isn't possible to tackle these problems in an emulator design where time is tied to the client's CPU.

**Comment by Tilka** [comment]
There is a patch for Quantum of Solace here: https://bugs.dolphin-emu.org/issues/9206#note-13

**Comment by jordan-woyak** [comment]
Here's something that I came up with for NFS: Carbon Gamecube USA (GW5E69)
It replaces the reentry of some kind of busy loop with a call to `VIWaitForRetrace`.
I wouldn't be surprised if something better is possible.
```
[OnFrame]
$Speed hack
0x8015E7D8:dword:0x48296E45
```

**Comment by SuperSamus** [comment]
Oh my, screw Excite Truck!
EDIT: Removed all the "documentation", since I'm pretty sure it was wrong.
But, in short, it seems that the game has its own synchronization mechanism, so the following patch removes it entirely in place of the gold old `VIWaitForRetrace`.
First line puts a `VIWaitForRetrace` after a `GXSetDrawDone`.
Second line removes an idle loop: the game's physics (which are fixed at 60 FPS with its own timer, running multiple times if the output frame rate is lower) have nothing to do with the vblank, meaning that there is still a wait after `VIWaitForRetrace`.
With this, the game's emulation performance doesn't decrease even at something absurd, like 10000% CPU Overclock.

```
[OnFrame]
$Limit internal FPS
0x802D2D00:dword:0x4BD87228
0x80177FBC:dword:0x60000000
```

Side effects:
- Raising VBI Frequency Override will speed up the game, which it didn't do before (this is caused by the second line).
- The game will now always run at a dividend of the vblank rate (meaning that it'll stay at a fixed 30 FPS during regular gameplay, without raising the Emulated CPU Clock), instead of being unstable. This could be fixable with a more complicated code that only conditionally calls `VIWaitForRetrace`, something like [this](https://discord.com/channels/521709831132807179/822820107788746812/1383061729226915883). (It technically would be an issue on all these games, it's just that this is the only one that doesn't run well beyond 60 FPS all the time.)
- Because the initial screen that tells you to hold the Wiimote horizontally does call `VIWaitForRetrace` (it still suffers from the idle loop), that screen will run at 30 FPS. Pretty sure no one will notice without an FPS monitor.

With this code, performance is improved* even at default clock rate: the first tutorial scenario goes from 72 to 84 FPS.
~~Oh, you wonder why the asterisk? Simple: _performance is only improved in single-core_. It becomes _worse_ in dual-core (at least it stays consistent when raising CPU clock, but still...).
Yes, I triple checked that.
Maybe the `VIWaitForRetrace` injection should be somewhere else?~~
Dual Core needs:
```
SyncOnSkipIdle = False
```
Otherwise the patch makes performance worse.

**Comment by SuperSamus** [comment]
Heck, technically the `SyncOnSkipIdle` note above applies to all games here: since Dolphin never detected an idle loop, then it never synced either, meaning that `SyncOnSkipIdle` didn't do anything on these games.
But with these patches, this setting now does something, and that may harm performance in dual core compared to not using the patch.

So... what shall be done?
- Add `SyncOnSkipIdle = False` to _all_ of these games (there is no accuracy regression by doing so, since they were already not syncing)
- Only add it on games where `SyncOnSkipIdle = True` + patch has a significant impact on performance (compared to being patchless on default CPU clock)

I vote for the former, because the potential performance regression of `SyncOnSkipIdle = True` is a regression.

**Comment by JMC47** [comment]
I'm fine with either.  On desktops, Dual Core is no longer the default so you're not going to magically make anything worse.  On Android, Sync on Idle is in the main UI and many users turn it off anyway, so if you prefer the former that's good with me.  These games aren't going to get worse, and Android + users enabling dualcore will want the performance in general.

**Comment by SuperSamus** [comment]
Added `SyncOnIdle = False` on these games.
Not added on the ones where the problem is only in specific scenarios (Bully and The Simpsons - Hit & Run).

Also, I found that Hulk... already had it. See #2569 for the historical reference.

**Review by JMC47 — APPROVED** [review]
I went through each INI to make sure everything was in the right place.  After going through twice, I'm pretty sure this should all just work.

There were some leftover requested changes from Rexogamer, but I couldn't actually find any unresolved conversations so I'm going to assume that's taken care of by now.

**Comment by SuperSamus** [comment]
Games that need to be tested because they *might* be in either list (I'll put the check when if tested and negative, or move them to the main list if positive):
- [Kuju Entertainment](https://wiki.dolphin-emu.org/index.php?title=Category:Kuju_Entertainment_(Developer))/[Headstrong Games](https://wiki.dolphin-emu.org/index.php?title=Category:Headstrong_Games_(Developer))
  - [ ] Battalion Wars (1 and 2)
  - [ ] Lord of the Rings: Aragorn's Quest
  - Others?
- [ ] 4x4 Evo 2 is positive, what about other games by [Terminal Reality](https://wiki.dolphin-emu.org/index.php?title=Category:Terminal_Reality_(Developer))?
- [ ] Deadly Creatures
- [ ] Marvel Nemesis: Rise of the Imperfects
- [ ] Flip's Twisted World
- [ ] Worms: Battle Islands
  - [ ] Worms 3D
  - [ ] Worms: A Space Oddity
  - [ ] Worms Blast
- [ ] X-Men Destiny, and by the same developers
  - [ ] Eternal Darkness: Sanity's Requiem
  - [ ] Metal Gear Solid: The Twin Snakes
- Other Need for Speed games
  - [ ] Hot Pursuit (2010) is a reskin of Nitro, so that's likely
  - [ ] The Run is a different engine
- Some of the [games that used to have the old kind of speed hack](https://github.com/dolphin-emu/dolphin/commit/cbe9c3e040a466bba410f025e026ed84f3f8b1c0) (Dolphin's idle detection used to be worse, so now they may be fine)

**Comment by kage2051** [comment]
Driver: San Francisco seems to run at an uncapped framerate during gameplay only, comic FMVs and the menu screen are not affected though.

**Comment by Adamillo** [comment]
Sonic Colors, Epic Mickey 2 and Donkey Kong Country Returns seemingly use a complex idle loop, as their performance gets lowered when the emulated CPU clock speed gets increased
